File: src/ee/apps_flyer/internal/AppsFlyerBridge.cpp
//
//  AppsFlyerBridge.cpp
//  ee_x_mobile_apps_flyer
//
//  Created by eps on 6/26/18.
//

#include "ee/apps_flyer/internal/AppsFlyerBridge.hpp"

#include <ee/nlohmann/json.hpp>

#include <ee/core/PluginManager.hpp>
#include <ee/core/Utils.hpp>
#include <ee/core/internal/MessageBridge.hpp>

namespace ee {
namespace core {
template <>
std::unique_ptr<IAppsFlyer> PluginManager::createPlugin() {
    return std::make_unique<apps_flyer::Bridge>();
}
} // namespace core

namespace apps_flyer {
namespace {
// clang-format off
const std::string kPrefix = "AppsFlyer";

const auto kInitialize      = kPrefix + "Initialize";
const auto kStartTracking   = kPrefix + "StartTracking";
const auto kGetDeviceId     = kPrefix + "GetDeviceId";
const auto kSetDebugEnabled = kPrefix + "SetDebugEnabled";
const auto kSetStopTracking = kPrefix + "SetStopTracking";
const auto kTrackEvent      = kPrefix + "TrackEvent";
// clang-format on
} // namespace

using Self = Bridge;

Self::Bridge()
    : bridge_(MessageBridge::getInstance()) {
    //
}

Self::~Bridge() {
    //
}

void Self::initialize(const std::string& devKey, const std::string& appId) {
    runOnUiThread([this, devKey, appId] {
#ifdef EE_X_ANDROID
        bridge_.call(kInitialize, devKey);
#else  // EE_X_ANDROID
        nlohmann::json json;
        json["devKey"] = devKey;
        json["appId"] = appId;
        bridge_.call(kInitialize, json.dump());
#endif // EE_X_ANDROID
    });
}

void Self::startTracking() {
    runOnUiThread([this] { //
        bridge_.call(kStartTracking);
    });
}

std::string Self::getDeviceId() const {
    return bridge_.call(kGetDeviceId);
}

void Self::setDebugEnabled(bool enabled) {
    runOnUiThread([this, enabled] {
        bridge_.call(kSetDebugEnabled, core::toString(enabled));
    });
}

void Self::setStopTracking(bool enabled) {
    runOnUiThread([this, enabled] {
        bridge_.call(kSetStopTracking, core::toString(enabled));
    });
}

void Self::trackEvent(const std::string& name,
                      const std::map<std::string, std::string>& values) {
    runOnUiThread([this, name, values] {
        nlohmann::json json;
        json["name"] = name;
        json["values"] = values;
        bridge_.call(kTrackEvent, json.dump());
    });
}
} // namespace apps_flyer
} // namespace ee

File: src/ee/apps_flyer/internal/AppsFlyerBridge.swift
//
//  AppsFlyerBridge.swift
//  ee-x-6914a733
//
//  Created by eps on 6/25/20.
//

import AppsFlyerLib

private let kPrefix = "AppsFlyerBridge"
private let kInitialize = "\(kPrefix)Initialize"
private let kStartTracking = "\(kPrefix)StartTracking"
private let kGetDeviceId = "\(kPrefix)GetDeviceId"
private let kSetDebugEnabled = "\(kPrefix)SetDebugEnabled"
private let kSetStopTracking = "\(kPrefix)SetStopTracking"
private let kTrackEvent = "\(kPrefix)TrackEvent"

@objc(EEAppsFlyerBridge)
public class AppsFlyerBridge: NSObject, IPlugin, AppsFlyerTrackerDelegate {
    private let _logger = Logger("\(AppsFlyerBridge.self)")
    private let _bridge: IMessageBridge
    private let _tracker = AppsFlyerTracker.shared()

    public required init(_ bridge: IMessageBridge) {
        _bridge = bridge
        super.init()
        registerHandlers()
    }

    public func destroy() {
        deregisterHandlers()
    }

    func registerHandlers() {
        _bridge.registerHandler(kInitialize) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let devKey = dict["devKey"] as? String else {
                assert(false, "Invalid argument")
            }
            guard let appId = dict["appId"] as? String else {
                assert(false, "Invalid argument")
            }
            self.intialize(devKey, appId)
            return ""
        }
        _bridge.registerHandler(kStartTracking) { _ in
            self.startTracking()
            return ""
        }
        _bridge.registerHandler(kGetDeviceId) { _ in
            self.deviceId
        }
        _bridge.registerHandler(kSetDebugEnabled) { message in
            self.setDebugEnabled(Utils.toBool(message))
            return ""
        }
        _bridge.registerHandler(kSetStopTracking) { message in
            self.setStopTracking(Utils.toBool(message))
            return ""
        }
        _bridge.registerHandler(kTrackEvent) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let name = dict["name"] as? String else {
                assert(false, "Invalid argument")
            }
            guard let values = dict["values"] as? [String: Any] else {
                assert(false, "Invalid argument")
            }
            self.trackEvent(name, values)
            return ""
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kStartTracking)
        _bridge.deregisterHandler(kGetDeviceId)
        _bridge.deregisterHandler(kSetDebugEnabled)
        _bridge.deregisterHandler(kSetStopTracking)
        _bridge.deregisterHandler(kTrackEvent)
    }

    func intialize(_ devKey: String, _ appId: String) {
        Thread.runOnMainThread {
            self._tracker.appsFlyerDevKey = devKey
            self._tracker.appleAppID = appId
            self._tracker.delegate = self
            self._tracker.shouldCollectDeviceName = true
            self._tracker.deviceTrackingDisabled = false
        }
    }

    func startTracking() {
        Thread.runOnMainThread {
            self._tracker.trackAppLaunch()
        }
    }

    var deviceId: String {
        return _tracker.getAppsFlyerUID()
    }

    func setDebugEnabled(_ enabled: Bool) {
        Thread.runOnMainThread {
            self._tracker.isDebug = enabled
        }
    }

    func setStopTracking(_ enabled: Bool) {
        Thread.runOnMainThread {
            self._tracker.isStopTracking = enabled
        }
    }

    func trackEvent(_ name: String, _ values: [String: Any]) {
        Thread.runOnMainThread {
            self._tracker.trackEvent(name, withValues: values)
        }
    }

    public func onConversionDataSuccess(_ conversionInfo: [AnyHashable: Any]) {
        _logger.debug("\(#function): \(conversionInfo)")
    }

    public func onConversionDataFail(_ error: Error) {
        _logger.debug("\(#function): \(error.localizedDescription)")
    }

    public func onAppOpenAttribution(_ attributionData: [AnyHashable: Any]) {
        _logger.debug("\(#function): \(attributionData)")
    }

    public func onAppOpenAttributionFailure(_ error: Error) {
        _logger.debug("\(#function): \(error.localizedDescription)")
    }

    public func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any]) -> Bool {
        _tracker.handleOpen(url, options: options)
        return false
    }

    public func application(_ application: UIApplication, open url: URL, sourceApplication: String?, annotation: Any) -> Bool {
        _tracker.handleOpen(url, sourceApplication: sourceApplication, withAnnotation: annotation)
        return false
    }

    public func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
        _tracker.continue(userActivity) { items in
            restorationHandler(items as? [UIUserActivityRestoring])
        }
        return false
    }
}
